site/content/entry/yuzu-progress-report-aug-2023/index.md

```diff
@@ -0,0 +1,531 @@
+Games use "queries" to fetch and load this data into memory whenever they need to use it.
+
+A Query Cache is a cache that emulates the GPU counters through the use of Host GPU queries. 
+yuzu's Query Cache runs these queries on the Host's (user) GPU to emulate the Switch's GPU counters.
+
+### What is it used for?
+
+Some games can make use of this fetched data to further optimise their rendering. For example, `SUPER MARIO ODYSSEY` optimises the amount of triangles it renders using a commonly used technique called "Occlusion Culling".
+
+Occlusion Culling is a feature that disables rendering of objects when they are not currently seen by the camera because they are obscured (occluded) by other objects.
+
+In `SUPER MARIO ODYSSEY`, the game first renders simple boundary boxes that cover all objects that will be rendered on the screen. 
+Then it'll count the pixels of each object using queries and later conditionally render the real objects based on whether any pixels of the object's corresponding box were rendered or not.
+
+Similarly, `Splatoon` games use pixel counts to check which characters are touching whose ink, if any at all. 
+The game tests for both ally ink and enemy ink and if both tests fail, then the character is not standing in any ink.
+
+{{< imgs
+	"./ink.png| Medium-well (Splatoon 2)"
+  >}}
+
+### Development & Challenges
+
+yuzu already has a Query Cache which was developed many years ago. 
+However, this implementation was not perfect and had many issues like not being able to hold any queries except pixel counts, not flushing everything in the correct order, not invalidating queries rewritten by other query types, among others.
+
+Our resident GPU dev [Blinkhawk](https://github.com/FernandoS27), set out to rework the Query Cache to fix these issues and modernise the code. 
+But, as he would later come to figure out, it wasn't going to be easy.
+
+If you recall, we mentioned earlier that a Query cache works by running queries on the host (user's) GPU. 
+Turns out this was actually the hardest part of the entire rewrite.
+
+During development, Blinkhawk ran into two big challenges with Host GPU queries. 
+The first issue was — when to do the queries and when to sync the memory?
+
+As soon as you start a game, the game begins making these queries to the GPU. 
+Normally, the results are obtained immediately after counting is done i.e., when the rendering is done. 
+But for games like `SUPER MARIO ODYSSEY`, that do Occlusion Culling, they make a lot of queries.
+
+{{< imgs
+	"./occ.png| Turn the camera around, and the whole city magically disappears (SUPER MARIO ODYSSEY)"
+  >}}
+
+So, if we tried to run all those queries and sync the results on the host (user's) GPU in the same way, it would stall the GPU heavily — as the GPU would be forced to draw each frame one by one while switching heavily between engines very often.
```

Comment:
```suggestion
So, if we tried to run all those queries and sync the results on the host (user's) GPU in the same way, it would stall the GPU heavily — as the GPU would be forced to synchronize with the CPU after drawing each scene element to write back the counters.
```